Torch/Utils/SteamWorkshopTools/SteamWorkshopTools.cs

```diff
@@ -0,0 +1,283 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using System.Xml.Serialization;
+using System.IO;
+using System.Net;
+using NLog;
+using SteamKit2;
+using System.Net.Http;
+
+namespace Torch.Utils.SteamWorkshopTools
+{
+    public class WebAPI
+    {
+        private static Logger Log = LogManager.GetLogger("SteamWorkshopService");
+        public const uint AppID = 244850U;
+        public string Username { get; private set; }
+        private string password;
+        public bool IsReady { get; private set; }
+        public bool IsRunning { get; private set; }
+        private TaskCompletionSource<bool> logonTaskCompletionSource;
+
+        private SteamClient steamClient;
+        private CallbackManager cbManager;
+        private SteamUser steamUser;
+
+        private static WebAPI _instance;
+        public static WebAPI Instance
+        {
+            get
+            {
+                return _instance ?? (_instance = new WebAPI());
+            }
+        }
+
+        private WebAPI()
+        {
+            steamClient = new SteamClient();
+            cbManager = new CallbackManager(steamClient);
+
+            IsRunning = true;
+        }
+
+        public async Task<bool> Logon(string user = "anonymous", string pw = "")
+        {
+            if (user == null)
+                throw new ArgumentNullException("User can't be null!");
+            if (user != "anonymous" && pw != "")
+                throw new ArgumentNullException("Password can't be null if user is not anonymous!");
+
+            Username = user;
+            password = pw;
+
+            logonTaskCompletionSource = new TaskCompletionSource<bool>();
+
+            steamUser = steamClient.GetHandler<SteamUser>();
+            cbManager.Subscribe<SteamClient.ConnectedCallback>(OnConnected);
+            cbManager.Subscribe<SteamClient.DisconnectedCallback>(OnDisconnected);
+            cbManager.Subscribe<SteamUser.LoggedOnCallback>(OnLoggedOn);
+            cbManager.Subscribe<SteamUser.LoggedOffCallback>(OnLoggedOff);
+
+            Log.Info("Connecting to Steam...");
+
+            steamClient.Connect();
+
+            await logonTaskCompletionSource.Task;
+            return logonTaskCompletionSource.Task.Result;
+        }
+
+        public void CancelLogon()
+        {
+            logonTaskCompletionSource?.SetCanceled();
+        }
+
+        public async Task<Dictionary<ulong, PublishedItemDetails>> GetPublishedFileDetails(IEnumerable<ulong> workshopIds)
+        {
+            //if (!IsReady)
+            //    throw new Exception("SteamWorkshopService not initialized!");
+
+            using (dynamic remoteStorage = SteamKit2.WebAPI.GetInterface("ISteamRemoteStorage"))
```

Comment:
Is Dynamic *absolutely* required here?

Reply:
I guess not, but it's how Steamkit2 is supposed (recommended) to be used.

Reply:
SteamKit2 is wrong :P Always prefer explicit types if they exist.

Reply:
Okay, before I go and change it, could you please have a quick look at Steamkit2's web API example and confirm that I should not use their recommended way of doing it?
The example I'm referring to is [this one](https://github.com/SteamRE/SteamKit/blob/master/Samples/6.WebAPI/Program.cs).

Unfortunately, these examples are the only source of documentation on Steamkit2 altogether (except the XML documentation comments and the source), so doing anything different is always a pain in the ass.

Reply:
Just saw, that there's an example for using the Steamkit WebAPI without dynamics at the bottom of the example. I'm going to use it that way then.

Reply:
Ah, I assumed there was a legitimate wrapper for the interface you were using. I guess `dynamic` is fine then since there's no "proper" C# way to invoke methods in that interface.

Reply:
Here's a side by side comparison:

statically typed:
```cs
var ifaceArgs = new Dictionary<string, string>();
ifaceArgs["itemcount"] = workshopIds.Count().ToString();
// no idea if that formatting is correct - in fact I get a 404 response
ifaceArgs["publishedfileids"] = string.Join(",", workshopIds); 
return remoteStorage.Call(HttpMethod.Post, "GetPublishedFileDetails", args:ifaceArgs);
```
dynamically typed:
```cs
return remoteStorage.GetPublishedFileDetails1(
    itemcount: workshopIds.Count(), 
    publishedfileids: workshopIds, 
    method: HttpMethod.Post);
```

Reply:
@Jimmacle Yeah, they would need to provide a Type for each SteamInterface in order to get any benefits from it.
Should I mark this as resolved then?